### DISCUSSION TITLE
Question: How to map log severity in GCP using Zap SugaredLogger

### DISCUSSION THREAD

**Discussion by gpkmr-genesys** [uber-go/zap#1110] [Q&A] [answered]
Hello - I have an application running on `GCP` `Cloud Functions`. Decided to use `Zap` `SugaredLogger` since the application is not intended for high performance.

Issue: All my logs are mapped in `default` logging level in GCP log explorer. Below is the sample code where I am using different log levels but still it fails to map. Unable to find any articles/examples for guidance. 

```
loggerConfig := zap.NewProductionConfig()
	loggerConfig.EncoderConfig.TimeKey = "ts"
	loggerConfig.EncoderConfig.EncodeTime = zapcore.TimeEncoderOfLayout(time.RFC3339)

	logger, err := loggerConfig.Build()
	if err != nil {
		log.Fatal(err)
	}

	defer logger.Sync()
	sugar = logger.Sugar()
        sugar.Debug("This is debug")
        sugar.Info("This is info")
        sugar.Warn("This is warning")
        sugar.Error("This is error")
```
The log entries are mapped to default level always.
![image](https://user-images.githubusercontent.com/76488975/167630452-b0c5e232-cef9-4be4-b0c0-a81fa1c8725a.png)

Below are the available log levels where * is the default.
![image](https://user-images.githubusercontent.com/76488975/167624455-ff3f4b3d-c6cf-40ba-9a3b-5975bd14f3d1.png)

GCP docs only has articles related to `cloud.google.com/go/logging` library. Since my application will be running on both GCP & Azure, I decided to go with common 3rd party logger - Zap.

It would be really helpful if someone can point me in the right direction. 

P.S: Am a beginner in Go and also using 3rd party loggers.

**Comment by vducong** [comment, marked as answer]
To log GCP severity correctly, first, you have to use JSON encoding, then map Zap's level with GCP severity and set appropriate keys with the LogEntry as [documented here](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry)

For example:
```go
func New(cfg *configs.Config) *Logger {
    loggerCfg := &zap.Config{
		Level:            zap.NewAtomicLevelAt(zapcore.InfoLevel),
		Encoding:         "json",
		EncoderConfig:    encoderConfig,
		OutputPaths:      []string{"stdout"},
		ErrorOutputPaths: []string{"stderr"},
    }

    plain, err := loggerCfg.Build(zap.AddStacktrace(zap.DPanicLevel))
    if err != nil {
		plain = zap.NewNop()
	}
	return plain.Sugar()
}


var encoderConfig = zapcore.EncoderConfig{
	TimeKey:        "time",
	LevelKey:       "severity",
	NameKey:        "logger",
	CallerKey:      "caller",
	MessageKey:     "message",
	StacktraceKey:  "stacktrace",
	LineEnding:     zapcore.DefaultLineEnding,
	EncodeLevel:    encodeLevel(),
	EncodeTime:     zapcore.RFC3339TimeEncoder,
	EncodeDuration: zapcore.MillisDurationEncoder,
	EncodeCaller:   zapcore.ShortCallerEncoder,
}

func encodeLevel() zapcore.LevelEncoder {
	return func(l zapcore.Level, enc zapcore.PrimitiveArrayEncoder) {
		switch l {
		case zapcore.DebugLevel:
			enc.AppendString("DEBUG")
		case zapcore.InfoLevel:
			enc.AppendString("INFO")
		case zapcore.WarnLevel:
			enc.AppendString("WARNING")
		case zapcore.ErrorLevel:
			enc.AppendString("ERROR")
		case zapcore.DPanicLevel:
			enc.AppendString("CRITICAL")
		case zapcore.PanicLevel:
			enc.AppendString("ALERT")
		case zapcore.FatalLevel:
			enc.AppendString("EMERGENCY")
		}
	}
}
```

**Reply by cybercent** [reply]
It appears that the argument `cfg *configs.Config` is not needed.

**Reply by vducong** [reply]
Sure. Configs are depended on your need and definitely not obligatory parameters.

**Reply by kevmo314** [reply]
`zapcore.RFC3339TimeEncoder` should be `zapcore.ISO8601TimeEncoder`. Cloud Logging doesn't correctly parse RFC3339 timestamps.

**Comment by juampynr** [comment]
Shall we close this issue?

**Reply by abhinav** [reply]
Thanks for bumping this, @juampynr!
I converted it into a discussion so others can find the answer easily in the future.
